fabtests/prov/efa/src/efa_mmap_test.c

```diff
@@ -0,0 +1,223 @@
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <sys/mman.h>
+#include <string.h>
+#include <errno.h>
+#include <shared.h>
+
+static void *mmap_buf;
+static struct fid_mr *mmap_mr;
+
+struct test_case {
+	const char *name;
+	int mmap_prot;
+	uint64_t mr_access;
+	bool should_pass;
+};
+
+static struct test_case test_cases[] = {
+	{"PROT_READ for FI_WRITE", PROT_READ, FI_WRITE, true},
+	{"PROT_READ for FI_REMOTE_WRITE", PROT_READ, FI_REMOTE_WRITE, false},
+	{"PROT_READ|PROT_WRITE for FI_REMOTE_WRITE", PROT_READ | PROT_WRITE, FI_REMOTE_WRITE, true},
+	{"PROT_READ for FI_RECV", PROT_READ, FI_RECV, false},
+	{"PROT_READ|PROT_WRITE for FI_RECV", PROT_READ | PROT_WRITE, FI_RECV, true},
+	{"PROT_READ for FI_READ", PROT_READ, FI_READ, false},
+	{"PROT_READ|PROT_WRITE for FI_READ", PROT_READ | PROT_WRITE, FI_READ, true},
+	{"PROT_READ for FI_REMOTE_READ", PROT_READ, FI_REMOTE_READ, true},
+};
+
+#define NUM_TEST_CASES (sizeof(test_cases) / sizeof(test_cases[0]))
+
+static int setup_mmap_buffer(int prot, uint64_t access)
+{
+	int ret;
+	void *mmap_desc;
+
+	/* Create writable mapping first to initialize data */
+	mmap_buf = mmap(NULL, opts.transfer_size, PROT_READ | PROT_WRITE,
+			MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+	if (mmap_buf == MAP_FAILED) {
+		FT_PRINTERR("mmap", -errno);
+		return -errno;
+	}
+
+	/* Initialize buffer with test data */
+	ft_fill_buf(mmap_buf, opts.transfer_size);
+
+	/* Set desired protection */
+	if (mprotect(mmap_buf, opts.transfer_size, prot) < 0) {
+		FT_PRINTERR("mprotect", -errno);
+		munmap(mmap_buf, opts.transfer_size);
+		return -errno;
+	}
+
+	/* Register memory region */
+	ret = ft_reg_mr(fi, mmap_buf, opts.transfer_size, access,
+			FT_MR_KEY, FI_HMEM_SYSTEM, 0, &mmap_mr, &mmap_desc);
+	if (ret) {
+		FT_PRINTERR("ft_reg_mr", ret);
+		munmap(mmap_buf, opts.transfer_size);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void cleanup_mmap_buffer(void)
+{
+	if (mmap_mr) {
+		fi_close(&mmap_mr->fid);
+		mmap_mr = NULL;
+	}
+	if (mmap_buf) {
+		munmap(mmap_buf, opts.transfer_size);
+		mmap_buf = NULL;
+	}
+}
+
+
+static int run_single_test(struct test_case *test)
+{
+	int ret;
+
+	printf("Testing: %s\n", test->name);
+
+	ret = setup_mmap_buffer(test->mmap_prot, test->mr_access);
+
+	if (ret) {
+		if (test->should_pass) {
+			printf("FAIL: %s - mr reg failed unexpectedly: %s\n\n", 
+			       test->name, fi_strerror(-ret));
+			cleanup_mmap_buffer();
+			return -1;
+		} else {
+			printf("PASS: %s - mr reg failed as expected\n\n", test->name);
```

Comment:
what if the failure is from the mprotect and mmap? Should have a way to distinguish them?